### ISSUE TITLE
DINOv2 is now available in HF Transformers (with tutorial)

### ISSUE THREAD

**Issue by NielsRogge** [facebookresearch/dinov2#153]
Hi folks,

As there are multiple issues here regarding fine-tuning DINOv2 on custom data, questions related to semantic segmentation/depth estimation, etc. I've created a [tutorial notebook](https://github.com/NielsRogge/Transformers-Tutorials/tree/master/DINOv2) on training a linear classifier using DINOv2's frozen features for semantic segmentation. The notebook would be very similar for image classification or depth estimation.

DINOv2 is now available in HF Transformers: https://huggingface.co/docs/transformers/main/model_doc/dinov2

The checkpoints are on the hub: https://huggingface.co/models?other=dinov2

Feature extraction is also very simple:

```
from transformers import AutoImageProcessor, AutoModel
from PIL import Image
import requests

url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
image = Image.open(requests.get(url, stream=True).raw)

processor = AutoImageProcessor.from_pretrained('facebook/dinov2-base')
model = AutoModel.from_pretrained('facebook/dinov2-base')

inputs = processor(images=image, return_tensors="pt")
outputs = model(**inputs)
features = outputs.last_hidden_state
```

Intermediate features can be easily obtained by passing `output_hidden_states=True` to the forward method.

Can be relevant for #6, #14, #15, #25, #46, #47, #54, #55, #80, #84, #97, #99

Have fun fine-tuning them!

Cheers,

Niels

**Comment by Starlento** [comment]
I have a silly question, sorry to ask here. For the hidden_states, I want to convert (batch_size, num_image_patches, embedding_dim) to (batch_size, h, w, embedding_dim) for segmentation tasks. But I found for a (224, 224) image, the num_image_patches is 257 (not 16x16=256). What is the correct way to reshape it?

**Comment by NielsRogge** [comment]
Hi @Starlento great question! This is because DINOv2 (and vision transformers in general) typically also add a special CLS token before the sequence of image patches. Hence the sequence length becomes (image_size/patch_size)**2 + 1. So in case you use a DINOv2 model with an image resolution of 224 and a patch size of 16, you get (224/16)**2 + 1 = 257 embeddings out.

Hence one usually discards the final embedding of the CLS token, and only uses the embeddings of the image patches, as done [here](https://github.com/facebookresearch/dinov2/blob/c3c2683a13cde94d4d99f523cf4170384b00c34c/dinov2/models/vision_transformer.py#L279). I think it makes sense to add a `Dinov2Backbone` class to the Transformers library, in similar spirit of other backbones. I've made a PR above for that.

Here's how you can use it (for now you'll need to do `pip install git+https://github.com/nielsrogge/transformers.git@add_dinov2_backbone`):

```
from transformers import Dinov2Backbone
import torch

model = Dinov2Backbone.from_pretrained("facebook/dinov2-base", out_indices=[0,1,2,3])

pixel_values = torch.randn(1, 3, 224, 224)

outputs = model(pixel_values)

for feature_map in outputs.feature_maps:
    print(feature_map.shape)
```
By default, feature maps will be 4D i.e. of shape (batch_size, num_channels, height, width). If you want 3D feature maps, just pass in `reshape=False` to the `from_pretrained` method.

**Comment by rainbowpuffpuff** [comment]
> I've created a [tutorial notebook](https://github.com/NielsRogge/Transformers-Tutorials/tree/master/DINOv2) on training a linear classifier using DINOv2's frozen features for semantic segmentation. The notebook would be very similar for image classification or depth estimation.


Hi, 

Thank you a lot for the tutorial and the hf version. I ran it on colab. I have two questions:

1) is the add_pooling_layer necessary? if set to false, will one pooling layer be added? because i don't think I came across this parameter in the class of the config/model from HF 


Error below: 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-37-433acb205ffe>](https://localhost:8080/#) in <cell line: 1>()
----> 1 model = Dinov2ForSemanticSegmentation.from_pretrained("facebook/dinov2-base", id2label=id2label, num_labels=len(id2label))

1 frames
[<ipython-input-36-e8c7a63af851>](https://localhost:8080/#) in __init__(self, config)
     23     super().__init__(config)
     24 
---> 25     self.dinov2 = Dinov2Model(config, add_pooling_layer=False)
     26     self.classifier = LinearClassifier(config.hidden_size, 32, 32, config.num_labels)
     27 

TypeError: Dinov2Model.__init__() got an unexpected keyword argument 'add_pooling_layer'




2) the rest of the code runs, but when I train the model on the EduardoPacheco/FoodSeg103 dataset (as it was in the code) 

it does train for a while but I guess it eventually encounters a datapoint it can't convert 

Error below: 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-35-c5ede5803258>](https://localhost:8080/#) in <cell line: 19>()
     19 for epoch in range(epochs):
     20   print("Epoch:", epoch)
---> 21   for idx, batch in enumerate(tqdm(train_dataloader)):
     22       pixel_values = batch["pixel_values"].to(device)
     23       labels = batch["labels"].to(device)

8 frames
[/usr/local/lib/python3.10/dist-packages/albumentations/core/composition.py](https://localhost:8080/#) in _check_args(self, **kwargs)
    284 
    285         if self.is_check_shapes and shapes and shapes.count(shapes[0]) != len(shapes):
--> 286             raise ValueError(
    287                 "Height and Width of image, mask or masks should be equal. You can disable shapes check "
    288                 "by setting a parameter is_check_shapes=False of Compose class (do it only if you are sure "

ValueError: Height and Width of image, mask or masks should be equal. You can disable shapes check by setting a parameter is_check_shapes=False of Compose class (do it only if you are sure about your data consistency).

**Comment by NielsRogge** [comment]
@rainbowpuffpuff thanks for reporting, I've removed `add_pooling_layer` recently, so no need to pass that. I've updated my notebook.

regarding the second question, looks like Albumentations says there's an image which has a segmentation mask with a different shape, weird, haven't encountered that. Will rerun the notebook to verify

**Comment by Starlento** [comment]
> Hi @Starlento great question! This is because DINOv2 (and vision transformers in general) typically also add a special CLS token before the sequence of image patches. Hence the sequence length becomes (image_size/patch_size)**2 + 1. So in case you use a DINOv2 model with an image resolution of 224 and a patch size of 16, you get (224/16)**2 + 1 = 257 embeddings out.
> 
> Hence one usually discards the final embedding of the CLS token, and only uses the embeddings of the image patches, as done [here](https://github.com/facebookresearch/dinov2/blob/c3c2683a13cde94d4d99f523cf4170384b00c34c/dinov2/models/vision_transformer.py#L279). I think it makes sense to add a `Dinov2Backbone` class to the Transformers library, in similar spirit of other backbones. I've made a PR above for that.
> 
> Here's how you can use it (for now you'll need to do `pip install git+https://github.com/nielsrogge/transformers.git@add_dinov2_backbone`):
> 
> ```
> from transformers import Dinov2Backbone
> import torch
> 
> model = Dinov2Backbone.from_pretrained("facebook/dinov2-base", out_indices=[0,1,2,3])
> 
> pixel_values = torch.randn(1, 3, 224, 224)
> 
> outputs = model(pixel_values)
> 
> for feature_map in outputs.feature_maps:
>     print(feature_map.shape)
> ```
> 
> By default, feature maps will be 4D i.e. of shape (batch_size, num_channels, height, width). If you want 3D feature maps, just pass in `reshape=False` to the `from_pretrained` method.

I found the reshape is somewhat wrong?
``` python
from transformers import Dinov2Backbone
import torch

encoder = Dinov2Backbone.from_pretrained("hf-base-models/facebook_dinov2-large", out_features=["stage6", "stage12", "stage18", "stage24"])
picked_hidden_states = encoder(torch.rand(1, 3, 448, 224)).feature_maps
for x in picked_hidden_states:
    print(x.shape)
torch.Size([1, 1024, 16, 32])
torch.Size([1, 1024, 16, 32])
torch.Size([1, 1024, 16, 32])
torch.Size([1, 1024, 16, 32])
```
I used to use only square images that I did not find the problem...

**Comment by NielsRogge** [comment]
Hi @Starlento,

That's indeed a bug in the original implementation, I'm addressed it in https://github.com/huggingface/transformers/pull/26092

**Comment by GLASS-z13** [comment]
does anybody see an example of using Mask2Former as a head?

**Comment by mfarre** [comment]
> is_check_shapes



> @rainbowpuffpuff thanks for reporting, I've removed `add_pooling_layer` recently, so no need to pass that. I've updated my notebook.
> 
> regarding the second question, looks like Albumentations says there's an image which has a segmentation mask with a different shape, weird, haven't encountered that. Will rerun the notebook to verify

Looks like at least one of the images in the dataset is transposed. As a quick hack, In the SegmentationDataset class, I added the following and with it, I could train:
    if original_image.shape[:2] != original_segmentation_map.shape:
       original_image = np.transpose(original_image, (1,0,2))
       print("Transposed and continuing")
       print("Original image "+str(original_image.shape))

**Comment by alvaro-stylesage** [comment]
> Hi folks,
> 
> As there are multiple issues here regarding fine-tuning DINOv2 on custom data, questions related to semantic segmentation/depth estimation, image similarity and feature extraction etc. this should now become easier given the model is available in HF Transformers. Check below for tips and tricks.
> 
> Documentation: https://huggingface.co/docs/transformers/main/model_doc/dinov2
> 
> The checkpoints are on the hub: https://huggingface.co/models?other=dinov2
> 
> I've created a [tutorial notebook](https://github.com/NielsRogge/Transformers-Tutorials/tree/master/DINOv2) on training a linear classifier using DINOv2's frozen features for semantic segmentation. The notebook would be very similar for image classification or depth estimation.
> ### Semantic segmentation/image classification/depth estimation
> 
> Refer to my demo notebook here: https://github.com/NielsRogge/Transformers-Tutorials/tree/master/DINOv2. One just places a linear classifier on top of the model, and uses the features as-is.
> 
> Depth estimation:
> 
>     * DPT + DINOv2 is now supported, a notebook has been made available [here](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/DPT/Inference_with_DPT_%2B_DINOv2_for_depth_estimation.ipynb).
> 
>     * For fine-tuning, one would however use a different loss function, like [this one](https://github.com/huggingface/transformers/blob/e42587f596181396e1c4b63660abf0c736b10dae/src/transformers/models/glpn/modeling_glpn.py#L765-L766) used in the [GLPN model](https://huggingface.co/docs/transformers/model_doc/glpn#transformers.GLPNForDepthEstimation) to predict the loss between logits and ground truth depth maps.
> 
> 
> Image classification: here the linear classifier can look simpler, you can just use `Dinov2ForImageClassification`. Refer to [this notebook](https://github.com/huggingface/notebooks/blob/main/examples/image_classification.ipynb) or [example scripts](https://github.com/huggingface/transformers/tree/main/examples/pytorch/image-classification).
> ### Feature extraction
> 
> Feature extraction is also very simple:
> 
> ```
> from transformers import AutoImageProcessor, AutoModel
> from PIL import Image
> import requests
> 
> url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
> image = Image.open(requests.get(url, stream=True).raw)
> 
> processor = AutoImageProcessor.from_pretrained('facebook/dinov2-base')
> model = AutoModel.from_pretrained('facebook/dinov2-base')
> 
> inputs = processor(images=image, return_tensors="pt")
> outputs = model(**inputs)
> features = outputs.last_hidden_state
> ```
> 
> The `features` in this case will be a PyTorch tensor of shape (batch_size, num_image_patches, embedding_dim). So one can turn them into a single vector by averaging over the image patches, like so:
> 
> ```
> features = features.mean(dim=1)
> ```
> 
> Now you have a single 768-dim (or other sizes, depending on which one you are using) vector for each image in your batch.
> ### Getting intermediate features
> 
> Intermediate features can be easily obtained by passing `output_hidden_states=True` to the forward method of the code snippet above. The `outputs` will then contain an additional key called `hidden_states`, which contain the intermediate features for each of the Transformer layers.
> ### Image similarity
> 
> We have a tutorial on that here: https://huggingface.co/blog/image-similarity. Given that DINOv2 now is available in HF Transformers, one can simply replace the model_ckpt in the blog with the [ones of DINOv2](https://huggingface.co/models?other=dinov2) on the 🤗 hub.
> 
> Can be relevant for #6, #14, #15, #25, #46, #47, #54, #55, #80, #84, #97, #99
> 
> Have fun fine-tuning them!
> 
> Cheers,
> 
> Niels

Hi @NielsRogge , thanks for the detailed information. Then, according to your explanation, doing:
```
outputs = model(**inputs)
outputs.last_hidden_state
```
should be exactly the same as doing:
```
outputs = model(**inputs, output_hidden_states=True)
outputs.hidden_states[-1]
```

In my case I find that these 2 chunks of code return completely different tensors, so I am not sure which one corresponds to the final CLS and patch embeddings at the end of the Transformer. If you could clarify it would be great.

Thanks!

**Comment by NielsRogge** [comment]
> In my case I find that these 2 chunks of code return completely different tensors, so I am not sure which one corresponds to the final CLS and patch embeddings at the end of the Transformer. If you could clarify it would be great.

Yes that's because there's layernorm applied on the last hidden states as seen [here](In my case I find that these 2 chunks of code return completely different tensors, so I am not sure which one corresponds to the final CLS and patch embeddings at the end of the Transformer. If you could clarify it would be great.

Thanks!).

**Comment by alvaro-stylesage** [comment]
> > In my case I find that these 2 chunks of code return completely different tensors, so I am not sure which one corresponds to the final CLS and patch embeddings at the end of the Transformer. If you could clarify it would be great.
> 
> Yes that's because there's layernorm applied on the last hidden states as seen [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/dinov2/modeling_dinov2.py#L643).
> 
> Thanks!).

Thanks for the answer. So when using DinoV2 as a feature extractor for images, is it better to take the embeddings after applying LayerNorm or before?

Thanks!

**Comment by NielsRogge** [comment]
It's mostly a matter of experimentation, I would just try out both and see which ones work best.

**Comment by edouardmercier** [comment]
@NielsRogge, just for the sake of clarify, to account for the output tensor of the second dimension `(image_size / 
 patch_size)^2 + 1`, from what I read and understood from the [model card](https://github.com/facebookresearch/dinov2/blob/main/MODEL_CARD.md), the model image patch size is of 14 pixels and not 16 pixels as you mentioned [here](https://github.com/facebookresearch/dinov2/issues/153#issuecomment-1678739255), so that with an image resolution of 224 and a patch size of 14, you get `(224 / 14)^2 + 1 = 257` embeddings out. Thank you so much for your work! ;)

**Comment by franchesoni** [comment]
Could you reproduce any of the paper's results?

**Comment by NielsRogge** [comment]
@franchesoni I ported the weigths to the HF format, to reproduce the results I'd recommend the scripts present in the original repository. We do have image classification scripts here: https://github.com/huggingface/transformers/tree/main/examples/pytorch/image-classification, but they are mostly for demo purposes, and need to be tweaked for a specific use case.

**Comment by zhaoyanpeng** [comment]
@NielsRogge thank you for your amazing contribution! please can you tell me more about the "[rescale_factor](https://huggingface.co/facebook/dinov2-base/blob/main/preprocessor_config.json#L23)" parameter, why do we need it? I did not find the corresponding piece of code from the official repo of dinov2, can you point it out?

**Comment by NielsRogge** [comment]
Hi @zhaoyanpeng that value comes from 1/255. Typically the red, green and blue color channels of images have values between 0 and 255. Neural networks on the other hand are typically trained on numbers between 0 and 1. So rescaling is a kind of standardization step. The original repository uses [ToTensor](https://pytorch.org/vision/stable/generated/torchvision.transforms.ToTensor.html) from torchvision which does the same thing.

**Comment by zhaoyanpeng** [comment]
> Hi @zhaoyanpeng that value comes from 1/255. Typically the red, green and blue color channels of images have values between 0 and 255. Neural networks on the other hand are typically trained on numbers between 0 and 1. So rescaling is a kind of standardization step. The original repository uses [ToTensor](https://pytorch.org/vision/stable/generated/torchvision.transforms.ToTensor.html) from torchvision which does the same thing.

Ah, it now makes sense. thnk you for your prompt reply!

**Comment by Yonggie** [comment]
Hi, does any have such issue like this?  ``key error dinov2``? It's with ``transformer==4.30.2`` and ``timm==0.9.12``
![图片](https://github.com/facebookresearch/dinov2/assets/31330044/43e3bac9-ebe2-413d-8bab-858b3a92d764)

**Comment by NielsRogge** [comment]
DINOv2 was probably added in a later version of Transformers, so `pip install --upgrade transformers` will fix that.

**Comment by Sundragon1993** [comment]
@NielsRogge thank you for the amazing works, I have a question regarding image similarity calculation, should I take the mean value of the last_hidden_state for each image in total of 2 images to compute
```
emb_img1, emb_img2 = last_hidden_states[0].mean(dim=0), last_hidden_states[1].mean(dim=0)
or
emb_img1, emb_img2 = last_hidden_states[0, 0], last_hidden_states[1, 0]  # Get cls token (0-th token) for each img
```

the second line of code successfully replicated the result of this paper:
https://openaccess.thecvf.com/content/CVPR2023/supplemental/Ruiz_DreamBooth_Fine_Tuning_CVPR_2023_supplemental.pdf
but the first line failed.

**Comment by NielsRogge** [comment]
Hi,

It depends a bit, some models have a CLS token which is specifically trained in a contrastive way like CLIP or SigLIP, so it's advised to use that. Other models work better by average pooling the final hidden state of the patch tokens.

So I would try both approaches and see which one works best.

**Comment by antopost** [comment]
Hi @NielsRogge 
What would be the necessary steps to use DINOv2 with ViT Adapter?
What would be the easiest way to achieve this?

**Comment by NielsRogge** [comment]
Update here, DINOv2 with Registers is now available!

Docs: https://huggingface.co/docs/transformers/main/en/model_doc/dinov2_with_registers
Checkpoints: https://huggingface.co/models?other=dinov2_with_registers

**Comment by MengHao666** [comment]
> https://github.com/NielsRogge/Transformers-Tutorials/tree/master/DINOv2

would u support the original dino v1?

**Comment by NielsRogge** [comment]
Yes the original DINO is supported too: https://huggingface.co/models?library=transformers&other=dino&sort=trending, see e.g. https://huggingface.co/facebook/dino-vitb16.

**Comment by NielsRogge** [comment]
Hi yes it's here: https://github.com/huggingface/transformers/blob/main/src/transformers/models/dinov2/convert_dinov2_to_hf.py
